packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -171,21 +194,79 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 			invalid: invalid || !!errorSummary,
 		});
 
+		const valueChangeToken = JSON.stringify(
+			value.filter((file) =>
+				fileRejections.every(
+					(rejectedFile) => rejectedFile.file.path !== file.path
+				)
+			)
+		);
+
 		useEffect(() => {
-			setFileRejections(
-				dropzoneFileRejections.map(({ file, errors }) => ({
-					file,
-					errors: errors.map((error) => ({
-						code: error.code,
-						message: getFileRejectionErrorMessage(
-							error,
-							formattedMaxFileSize,
-							acceptedFilesSummary
-						),
-					})),
-				}))
-			);
-		}, [dropzoneFileRejections, formattedMaxFileSize, acceptedFilesSummary]);
+			const allAttemptedUploads = [
+				...value,
+				...fileRejections.map((rej) => rej.file),
+			];
+			const isOverMaxFilesLimit =
+				maxFiles && allAttemptedUploads.length > maxFiles;
```

Comment:
Could you add the `length`s of `value` and `fileRejections` rather than the spreads and `map`?

```suggestion
			const isOverMaxFilesLimit =
				maxFiles && value.length + fileRejections.length > maxFiles;
```

Reply:
I was using `allAttemptedUploads` at one point as a list of files however I ended up moving away from using it. I guess I didn't end up using it for anything other than a total files count.